File: js/packages/core/src/diagnostic.js
/**
 * Diagnostic codes. These are stable API: the conformance corpus asserts them,
 * and messages are deliberately not asserted (AGENTS.md §9).
 */
export const codes = {
  inputTooLong: 'input_too_long',
  unterminatedString: 'unterminated_string',
  invalidEscape: 'invalid_escape',
  unexpectedToken: 'unexpected_token',
  unexpectedEOF: 'unexpected_eof',
  unbalancedParen: 'unbalanced_paren',
  depthExceeded: 'depth_exceeded',
  tooManyPredicates: 'too_many_predicates',
  unknownField: 'unknown_field',
  unknownOperatorForField: 'unknown_operator_for_field',
  invalidValueForField: 'invalid_value_for_field',
  invalidDuration: 'invalid_duration',
  unknownSortKey: 'unknown_sort_key',
  schemaInvalid: 'schema_invalid',
}

/**
 * A diagnostic is `{ code, message, span: [start, end], suggestions? }`, where
 * the span is a half-open range of 0-based codepoint offsets.
 */
export function diagnostic(code, span, message, suggestions) {
  const d = { code, message, span }
  if (suggestions && suggestions.length > 0) d.suggestions = suggestions
  return d
}

/**
 * Orders diagnostics by position, so "the first diagnostic" means the leftmost
 * problem regardless of which stage found it.
 */
export function sortDiagnostics(diagnostics) {
  return diagnostics.sort((a, b) =>
    a.span[0] !== b.span[0] ? a.span[0] - b.span[0] : a.span[1] - b.span[1],
  )
}

/** Thrown by compile and orderBy; carries the first diagnostic. */
export class SluiceError extends Error {
  constructor(diag) {
    super(`${diag.code}${diag.message ? `: ${diag.message}` : ''}`)
    this.name = 'SluiceError'
    this.diagnostic = diag
  }
}

/** Thrown when a schema fails AGENTS.md §4 validation. Carries every problem. */
export class SchemaError extends Error {
  constructor(diagnostics) {
    super(`invalid schema: ${diagnostics.map((d) => d.message).join('; ')}`)
    this.name = 'SchemaError'
    this.diagnostics = diagnostics
  }
}

File: ast/diagnostic.go
package ast

import (
	"encoding/json"
	"errors"
	"strconv"
)

// Diagnostic codes. These are stable API: the conformance corpus asserts them,
// and messages are deliberately not asserted (AGENTS.md §9).
const (
	CodeInputTooLong            = "input_too_long"
	CodeUnterminatedString      = "unterminated_string"
	CodeInvalidEscape           = "invalid_escape"
	CodeUnexpectedToken         = "unexpected_token"
	CodeUnexpectedEOF           = "unexpected_eof"
	CodeUnbalancedParen         = "unbalanced_paren"
	CodeDepthExceeded           = "depth_exceeded"
	CodeTooManyPredicates       = "too_many_predicates"
	CodeUnknownField            = "unknown_field"
	CodeUnknownOperatorForField = "unknown_operator_for_field"
	CodeInvalidValueForField    = "invalid_value_for_field"
	CodeInvalidDuration         = "invalid_duration"
	CodeUnknownSortKey          = "unknown_sort_key"
	CodeSchemaInvalid           = "schema_invalid"
)

// Span is a half-open [Start, End) range of 0-based Unicode codepoint offsets
// into the input (AGENTS.md §3). It marshals as a two-element JSON array.
type Span struct {
	Start int
	End   int
}

// MarshalJSON renders the span as [start, end].
func (s Span) MarshalJSON() ([]byte, error) {
	b := make([]byte, 0, 16)
	b = append(b, '[')
	b = strconv.AppendInt(b, int64(s.Start), 10)
	b = append(b, ',')
	b = strconv.AppendInt(b, int64(s.End), 10)
	b = append(b, ']')
	return b, nil
}

// UnmarshalJSON accepts the [start, end] form.
func (s *Span) UnmarshalJSON(data []byte) error {
	var pair []int
	if err := json.Unmarshal(data, &pair); err != nil {
		return err
	}
	if len(pair) != 2 {
		return errors.New("span must be a two-element array")
	}
	s.Start, s.End = pair[0], pair[1]
	return nil
}

// Diagnostic is a single problem found in the input, positioned in it.
// Messages are for display and are free to change; Code and Span are not.
type Diagnostic struct {
	Code        string   `json:"code"`
	Message     string   `json:"message,omitempty"`
	Span        Span     `json:"span"`
	Suggestions []string `json:"suggestions,omitempty"`
}
